core/src/Cuda/Kokkos_Cuda_Half.hpp

```diff
@@ -683,6 +746,92 @@ class half_t {
     return __half2float(val) >= __half2float(rhs.val);
 #endif
   }
+
+  KOKKOS_FUNCTION
+  friend bool operator==(const volatile half_t& lhs,
+                         const volatile half_t& rhs) {
+#ifdef __CUDA_ARCH__
+    return static_cast<bool>(const_cast<const impl_type&>(lhs.val) ==
+                             const_cast<const impl_type&>(rhs.val));
```

Comment:
I believe this won't provide the semantics that the signature promises, and worse results in undefined behavior:
"referring to a volatile object through a non-volatile glvalue results in undefined behavior. " https://en.cppreference.com/w/cpp/language/const_cast

Rather, the body of the `__CUDA_ARCH__` branch of this and later functions could be written as

```c++
impl_type lv = lhs.val;
impl_type rv = rhs.val;
return lv == rv;
```

That way the member accesses are actually volatile

Reply:
Agreed. This seems so simple now... 🤦

Thank you!!